libamqpprox/amqpprox_connectionlimitermanager.cpp

```diff
@@ -0,0 +1,266 @@
+/*
+** Copyright 2022 Bloomberg Finance L.P.
+**
+** Licensed under the Apache License, Version 2.0 (the "License");
+** you may not use this file except in compliance with the License.
+** You may obtain a copy of the License at
+**
+**     http://www.apache.org/licenses/LICENSE-2.0
+**
+** Unless required by applicable law or agreed to in writing, software
+** distributed under the License is distributed on an "AS IS" BASIS,
+** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+** See the License for the specific language governing permissions and
+** limitations under the License.
+*/
+
+#include <amqpprox_connectionlimitermanager.h>
+
+#include <amqpprox_connectionlimiterinterface.h>
+#include <amqpprox_fixedwindowconnectionratelimiter.h>
+#include <amqpprox_logging.h>
+
+#include <memory>
+#include <string>
+
+namespace Bloomberg {
+namespace amqpprox {
+
+ConnectionLimiterManager::ConnectionLimiterManager()
+: d_connectionRateLimitersPerVhost()
+, d_alarmOnlyConnectionRateLimitersPerVhost()
+, d_defaultConnectionRateLimit(0)
+, d_defaultAlarmOnlyConnectionRateLimit(0)
+, d_mutex()
+{
+}
+
+std::shared_ptr<ConnectionLimiterInterface>
+ConnectionLimiterManager::addConnectionRateLimiter(
+    const std::string &vhostName,
+    uint32_t           numberOfConnections)
+{
+    std::lock_guard<std::mutex> lg(d_mutex);
+
+    std::shared_ptr<FixedWindowConnectionRateLimiter> connectionRateLimiter =
+        std::make_shared<FixedWindowConnectionRateLimiter>(
+            numberOfConnections);
+    d_connectionRateLimitersPerVhost[vhostName] = {true,
```

Comment:
do we care if we're splatting over a previous entry? perhaps doing an `find/insert` would be better here? and logging what we splatted over?

Reply:
Yes you are right. We don't care about the previous entry. We can put the log to mention about the old entry. But that will increase the number of instructions inside synchronous code block.